### PR TITLE
chore: polars type hinting

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [x] 🐳 Other

## If you have comments or can explain your changes, please do so below

Similar to #1465 for polars module. There are a few stuff I could not manage to annotate, or better I was but then the magical `_expression_parsing.py` module was yelling _a lot_

### CODE REVIEW

File: narwhals/_polars/expr.py [hunk 1]
```diff
@@ -181,11 +181,11 @@ def name(self: Self) -> PolarsExprNameNamespace:
...
 
 class PolarsExprDateTimeNamespace:
-    def __init__(self, expr: PolarsExpr) -> None:
+    def __init__(self: Self, expr: PolarsExpr) -> None:
         self._expr = expr
 
-    def __getattr__(self, attr: str) -> Callable[[Any], PolarsExpr]:
-        def func(*args: Any, **kwargs: Any) -> PolarsExpr:
+    def __getattr__(self: Self, attr: str) -> Any:
+        def func(*args: Any, **kwargs: Any) -> Any:
```
Comment (MarcoGorelli) [inline narwhals/_polars/expr.py:188]: are the `Callable` / `PolarsExpr` types not correct?

**Review by MarcoGorelli — COMMENTED** [review]
nice! just got a question but the rest looks good

File: narwhals/_polars/expr.py (hunk 1 above)
Comment (FBruzzesi) [reply narwhals/_polars/expr.py:188]: Whops! let me roll back these! For some reason, I started to have issues at a certain point. I will double check

File: narwhals/_polars/expr.py [hunk 2]
```diff
@@ -20,22 +20,22 @@
...
         return "PolarsExpr"
 
-    def _from_native_expr(self, expr: pl.Expr) -> Self:
+    def _from_native_expr(self: Self, expr: pl.Expr) -> Self:
         return self.__class__(
             expr, dtypes=self._dtypes, backend_version=self._backend_version
         )
 
-    def __getattr__(self, attr: str) -> Any:
+    def __getattr__(self: Self, attr: str) -> Any:
```
Comment (FBruzzesi) [inline narwhals/_polars/expr.py:38]: @MarcoGorelli this was the critical one! The other `__getattr__` are fine, sorry for that

**Review by MarcoGorelli — APPROVED** [review]
thanks @FBruzzesi !

**Pull request merged**